// @picovoice/porcupine-react-native/lib/module/builtin_keywords.js
//
// Copyright 2021 Picovoice Inc.
//
// You may not use this file except in compliance with the license. A copy of the license is located in the "LICENSE"
// file accompanying this source.
//
// Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
// an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
// specific language governing permissions and limitations under the License.
//
var BuiltInKeywords = /*#__PURE__*/function (BuiltInKeywords) {
  BuiltInKeywords["ALEXA"] = "alexa";
  BuiltInKeywords["AMERICANO"] = "americano";
  BuiltInKeywords["BLUEBERRY"] = "blueberry";
  BuiltInKeywords["BUMBLEBEE"] = "bumblebee";
  BuiltInKeywords["COMPUTER"] = "computer";
  BuiltInKeywords["GRAPEFRUIT"] = "grapefruit";
  BuiltInKeywords["GRASSHOPPER"] = "grasshopper";
  BuiltInKeywords["HEY_GOOGLE"] = "hey google";
  BuiltInKeywords["HEY_SIRI"] = "hey siri";
  BuiltInKeywords["JARVIS"] = "jarvis";
  BuiltInKeywords["OK_GOOGLE"] = "ok google";
  BuiltInKeywords["PICOVOICE"] = "picovoice";
  BuiltInKeywords["PORCUPINE"] = "porcupine";
  BuiltInKeywords["TERMINATOR"] = "terminator";
  return BuiltInKeywords;
}(BuiltInKeywords || {});
export default BuiltInKeywords;
