// @picovoice/porcupine-react-native/lib/commonjs/porcupine_errors.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.PorcupineStopIterationError = exports.PorcupineRuntimeError = exports.PorcupineMemoryError = exports.PorcupineKeyError = exports.PorcupineInvalidStateError = exports.PorcupineInvalidArgumentError = exports.PorcupineIOError = exports.PorcupineError = exports.PorcupineActivationThrottledError = exports.PorcupineActivationRefusedError = exports.PorcupineActivationLimitError = exports.PorcupineActivationError = void 0;
//
// Copyright 2021 Picovoice Inc.
//
// You may not use this file except in compliance with the license. A copy of the license is located in the "LICENSE"
// file accompanying this source.
//
// Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
// an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
// specific language governing permissions and limitations under the License.
//

class PorcupineError extends Error {
  constructor(message) {
    super(message);
    this.name = 'PorcupineError';
  }
}
exports.PorcupineError = PorcupineError;
class PorcupineMemoryError extends PorcupineError {
  constructor(message) {
    super(message);
    this.name = 'PorcupineMemoryError';
  }
}
exports.PorcupineMemoryError = PorcupineMemoryError;
class PorcupineIOError extends PorcupineError {
  constructor(message) {
    super(message);
    this.name = 'PorcupineIOError';
  }
}
exports.PorcupineIOError = PorcupineIOError;
class PorcupineInvalidArgumentError extends PorcupineError {
  constructor(message) {
    super(message);
    this.name = 'PorcupineInvalidArgumentError';
  }
}
exports.PorcupineInvalidArgumentError = PorcupineInvalidArgumentError;
class PorcupineStopIterationError extends PorcupineError {
  constructor(message) {
    super(message);
    this.name = 'PorcupineStopIterationError';
  }
}
exports.PorcupineStopIterationError = PorcupineStopIterationError;
class PorcupineKeyError extends PorcupineError {
  constructor(message) {
    super(message);
    this.name = 'PorcupineKeyError';
  }
}
exports.PorcupineKeyError = PorcupineKeyError;
class PorcupineInvalidStateError extends PorcupineError {
  constructor(message) {
    super(message);
    this.name = 'PorcupineInvalidStateError';
  }
}
exports.PorcupineInvalidStateError = PorcupineInvalidStateError;
class PorcupineRuntimeError extends PorcupineError {
  constructor(message) {
    super(message);
    this.name = 'PorcupineRuntimeError';
  }
}
exports.PorcupineRuntimeError = PorcupineRuntimeError;
class PorcupineActivationError extends PorcupineError {
  constructor(message) {
    super(message);
    this.name = 'PorcupineActivationError';
  }
}
exports.PorcupineActivationError = PorcupineActivationError;
class PorcupineActivationLimitError extends PorcupineError {
  constructor(message) {
    super(message);
    this.name = 'PorcupineActivationLimitError';
  }
}
exports.PorcupineActivationLimitError = PorcupineActivationLimitError;
class PorcupineActivationThrottledError extends PorcupineError {
  constructor(message) {
    super(message);
    this.name = 'PorcupineActivationThrottledError';
  }
}
exports.PorcupineActivationThrottledError = PorcupineActivationThrottledError;
class PorcupineActivationRefusedError extends PorcupineError {
  constructor(message) {
    super(message);
    this.name = 'PorcupineActivationRefusedError';
  }
}
exports.PorcupineActivationRefusedError = PorcupineActivationRefusedError;
